### PR TITLE
sound: use descriptor_utils.rs to manipulate requests

### PR DESCRIPTION
**Pull request by MatiasVara**
### Summary of the PR

This commit makes `device.rs` to use `descriptor_utils.rs` API to read/write requests independent of the descriptor distribution. This is a draft PR in which I am still working on.

Thanks for the comments.

### Requirements

Before submitting your PR, please make sure you addressed the following
requirements:

- [x] All commits in this PR are signed (with `git commit -s`), and the commit
  message has max 60 characters for the summary and max 75 characters for each
  description line.
- [x] All added/changed functionality has a corresponding unit/integration
  test.
- [x] All added/changed public-facing functionality has entries in the "Upcoming 
  Release" section of CHANGELOG.md (if no such section exists, please create one).
- [x] Any newly added `unsafe` code is properly documented.

### CODE REVIEW

File: vhost-device-sound/src/device.rs [hunk 1]
```diff
@@ -301,21 +253,19 @@ impl VhostUserSoundThread {
                             p.direction = s.direction as u8;
                             p.channels_min = s.channels_min;
                             p.channels_max = s.channels_max;
-                            buf.extend_from_slice(p.as_slice());
+                            if let Err(err) = writter_content.write_obj(p) {
+                                log::error!("Error: DescriptorWriteFailed: {}", err);
```
Comment (Ablu) [inline vhost-device-sound/src/device.rs:257]: Is this new log entry intended?

Comment (MatiasVara) [reply vhost-device-sound/src/device.rs:257]: I tried to replicate what happened before, e.g., https://github.com/rust-vmm/vhost-device/blob/e19e93055a047d1fb11ad53b149bd11db9abcc8c/vhost-device-sound/src/device.rs#L310, is that OK?

**Comment by MatiasVara** [comment]
@Ablu @epilys feel free to comment/review. I'll keep it as a draft until virtio-queue is released. Thanks!

File: vhost-device-sound/src/device.rs [hunk 2]
```diff
@@ -155,137 +155,89 @@ impl VhostUserSoundThread {
...
+            let mem = atomic_mem.memory();
 
-            // Request descriptor.
-            let desc_request = descriptors[0];
-            if desc_request.is_write_only() {
-                return Err(Error::UnexpectedWriteOnlyDescriptor(0).into());
-            }
+            let Ok(mut reader) = desc_chain.clone().reader(&mem) else {
+                return Err(Error::DescriptorReadFailed.into());
+            };
```
Comment (Ablu) [inline vhost-device-sound/src/device.rs:161]: The else branch could be a `.map_err(...)?`, right? Maybe it would make sense to bubble up the internal error too by wrapping it in `DescriptorReadFailed`

same for later code

File: vhost-device-sound/src/device.rs [hunk 3]
```diff
@@ -155,137 +155,89 @@ impl VhostUserSoundThread {
...
                     if chmaps.len() <= start_id || chmaps.len() < start_id + count {
                         resp.code = VIRTIO_SND_S_BAD_MSG.into();
                     } else {
-                        let desc_response = descriptors[2];
-                        let mut buf = vec![];
-
                         for i in chmaps.iter().skip(start_id).take(count) {
-                            buf.extend_from_slice(i.as_slice());
+                            if let Err(err) = writter_content.write_obj(*i) {
+                                log::error!("Error: DescriptorWriteFailed: {}", err);
```
Comment (Ablu) [inline vhost-device-sound/src/device.rs:198]: was this new log line intended?

File: vhost-device-sound/src/device.rs [hunk 4]
```diff
@@ -155,137 +155,89 @@ impl VhostUserSoundThread {
...
                     if jacks.len() <= start_id || jacks.len() < start_id + count {
                         resp.code = VIRTIO_SND_S_BAD_MSG.into();
                     } else {
-                        let desc_response = descriptors[2];
-                        let mut buf = vec![];
-
                         for i in jacks.iter().skip(start_id).take(count) {
-                            buf.extend_from_slice(i.as_slice());
+                            if let Err(err) = writter_content.write_obj(*i) {
+                                log::error!("Error: DescriptorWriteFailed: {}", err);
```
Comment (Ablu) [inline vhost-device-sound/src/device.rs:219]: same here?

File: vhost-device-sound/src/device.rs (hunk 1 above)
Comment (Ablu) [reply vhost-device-sound/src/device.rs:256]: That did not do logging itself but just bubbled up the error. I assume the caller of this function then eventually logs it (or bubbles it up further). The new code returns the same error, so if the previous code already logged up in the call stack, this adds a second log before that.

File: vhost-device-sound/src/device.rs [hunk 5]
```diff
@@ -984,8 +934,13 @@ mod tests {
...
             let descs = [
-                Descriptor::new(addr_req, 0x100, 0, 0), // request
-                Descriptor::new(0x20_0000, 0x100, VRING_DESC_F_WRITE as u16, 0), // response
+                Descriptor::new(addr_req, size_of::<VirtioSoundHeader>() as u32, 0, 0), // request
+                Descriptor::new(
+                    0x20_0000,
+                    size_of::<VirtioSoundHeader>() as u32,
+                    VRING_DESC_F_WRITE as u16,
+                    0,
+                ), // response
```
Comment (Ablu) [inline vhost-device-sound/src/device.rs:887]: This probably could go into a separate commit? It does not seem to be related to the rest of the changes that happen here, or is there a relation?

File: vhost-device-sound/src/stream.rs [hunk 6]
```diff
@@ -239,85 +244,92 @@ impl Default for PcmParams {
...
+
+        Ok(bytes_written as u32)
     }
 
     #[inline]
     /// Returns the length of the sound data [`virtio_queue::Descriptor`].
-    pub fn desc_len(&self) -> u32 {
-        self.data_descriptor.len()
+    pub const fn len(&self) -> u32 {
+        self.len as u32
```
Comment (Ablu) [inline vhost-device-sound/src/stream.rs:320]: Any reason this needs to return an u32? This seems like it could potentially overflow?

File: vhost-device-sound/src/stream.rs [hunk 7]
```diff
@@ -362,7 +376,7 @@ mod tests {
 
         let desc_out = Descriptor::new(
             next_addr,
-            std::mem::size_of::<R>() as u32,
+            std::mem::size_of::<R>() as u32 + payload_len,
```
Comment (Ablu) [inline vhost-device-sound/src/stream.rs:379]: I did not check, but I assume the Descriptor takes a usize/u64 here? So maybe this should both convert to u64 before summing to avoid overflows?

File: vhost-device-sound/src/audio_backends/alsa.rs [hunk 8]
```diff
@@ -196,21 +196,21 @@ fn write_samples_direct(
     mmap: &mut alsa::direct::pcm::MmapPlayback<u8>,
 ) -> AResult<bool> {
     while mmap.avail() > 0 {
-        let Some(buffer) = stream.buffers.front_mut() else {
+        let Some(request) = stream.requests.front_mut() else {
```
Comment (Ablu) [inline vhost-device-sound/src/audio_backends/alsa.rs:199]: The renaming merged with the other refactoring makes it a bit harder than necessary to review the actual functional changes. IMHO it is still manageable here, but for the future you might want to consider doing that in a separate commit.

File: vhost-device-sound/src/stream.rs [hunk 9]
```diff
@@ -239,85 +244,92 @@ impl Default for PcmParams {
     }
 }
 
-pub struct Buffer {
-    data_descriptor: virtio_queue::Descriptor,
+pub struct Request {
+    //data_descriptor: virtio_queue::Descriptor,
```
Comment (Ablu) [inline vhost-device-sound/src/stream.rs:248]: why is this commented out? It is not described in the commit message of this commit.

File: vhost-device-sound/src/stream.rs [hunk 10]
```diff
@@ -239,85 +244,92 @@ impl Default for PcmParams {
...
     direction: Direction,
 }
 
-impl std::fmt::Debug for Buffer {
+impl std::fmt::Debug for Request {
     fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
-        fmt.debug_struct(stringify!(Buffer))
+        fmt.debug_struct(stringify!(Request))
+            .field("len", &self.len)
             .field("pos", &self.pos)
```
Comment (Ablu) [inline vhost-device-sound/src/stream.rs:259]: I would try to keep these ordered the same way as they are in the definition. That is of course violated already, but maybe we should not scramble it further? :)

File: vhost-device-sound/Cargo.toml [hunk 11]
```diff
@@ -36,5 +36,5 @@ pw = { package = "pipewire", version = "0.8", optional = true }
 [dev-dependencies]
 rstest = "0.18.2"
 tempfile = "3.9"
-virtio-queue = { version = "0.11", features = ["test-utils"] }
+virtio-queue = { version = "0.11", git = "https://github.com/rust-vmm/vm-virtio.git", features = ["test-utils"]}
```
Comment (Ablu) [inline vhost-device-sound/Cargo.toml:39]: I understand that this is still provisional, but why does this commit come last? I would have expected it to come first in history.

**Review by Ablu — COMMENTED** [review]
did a quick pass, mostly the `let ... else {return Err(...)}` pattern confuses me a bit. I think so far we mostly just used `let foo = (...).map_err(Error::WrappingErrror)?`

File: vhost-device-sound/src/device.rs (hunk 5 above)
Comment (MatiasVara) [reply vhost-device-sound/src/device.rs:887]: There is a relation. The test is checking for failure when a malformed control request has arrived, i.e., `ControlMessageKind::PcmInfo`, `ControlMessageKind::ChmapInfo`, `ControlMessageKind::JackInfo`. In the previous code, this required three descriptors so the test was creating two descriptors to check if the processing of the requests effectively fail. With the new code, the distribution over descriptors does not matter and the test was failing. To fix this, I had to reduce the len of the second descriptor to a value less than the expected payload for control msg above. Fixing this is somehow related with the commit. I do not know if I fixed it in the right way.

File: vhost-device-sound/src/audio_backends/alsa.rs (hunk 8 above)
Comment (MatiasVara) [reply vhost-device-sound/src/audio_backends/alsa.rs:199]: I agree. The reason I used buffer instead of request was that, with this commit, the entire request is enqueued instead of individual buffers. I thought the variable name should reflect the changed meaning of the queue elements. Do you think it would be better to keep buffer?

File: vhost-device-sound/src/stream.rs (hunk 9 above)
Comment (MatiasVara) [reply vhost-device-sound/src/stream.rs:248]: I forgot to remove this comment. Apologies. This field is not require since we can directly use the DescriptorChain from the IOMessage.

File: vhost-device-sound/src/stream.rs (hunk 6 above)
Comment (MatiasVara) [reply vhost-device-sound/src/stream.rs:320]: I was thinking to keep it as u32 since I use it for `used_len` in https://github.com/MatiasVara/vhost-device/blob/3b97c14d9457d838c8b44d64e3c3c48b30da47c3/vhost-device-sound/src/stream.rs#L343

File: vhost-device-sound/src/stream.rs (hunk 7 above)
Comment (MatiasVara) [reply vhost-device-sound/src/stream.rs:379]: Descriptor takes u32 as len.

File: vhost-device-sound/Cargo.toml (hunk 11 above)
Comment (MatiasVara) [reply vhost-device-sound/Cargo.toml:39]: I will move it to the beginning.

File: vhost-device-sound/src/stream.rs (hunk 6 above)
Comment (MatiasVara) [reply vhost-device-sound/src/stream.rs:319]: I wonder if I should make `self.len` also u32. At the moment, it is usize.

File: vhost-device-sound/src/device.rs [hunk 12]
```diff
@@ -155,86 +155,58 @@ impl VhostUserSoundThread {
...
 
-            // Request descriptor.
-            let desc_request = descriptors[0];
-            if desc_request.is_write_only() {
-                return Err(Error::UnexpectedWriteOnlyDescriptor(0).into());
-            }
+            let mut reader = desc_chain
+                .clone()
+                .reader(&mem)
+                .map_err(|_| Error::DescriptorReadFailed)?;
```
Comment (stefano-garzarella) [inline vhost-device-sound/src/device.rs:162]: Maybe here we can reuse the `UnexpectedWriteOnlyDescriptor` error.

File: vhost-device-sound/src/device.rs [hunk 13]
```diff
@@ -416,13 +361,16 @@ impl VhostUserSoundThread {
...
+            writter_status
+                .write_obj(resp)
                 .map_err(|_| Error::DescriptorWriteFailed)?;
-            used_len += desc_hdr.len();
 
-            if vring.add_used(desc_chain.head_index(), used_len).is_err() {
+            used_len += writter_status.bytes_written();
+
+            if vring
+                .add_used(desc_chain.head_index(), used_len as u32)
```
Comment (stefano-garzarella) [inline vhost-device-sound/src/device.rs:370]: I think we should at least log if used_len is overflow

File: vhost-device-sound/src/device.rs (hunk 5 above)
Comment (stefano-garzarella) [reply vhost-device-sound/src/device.rs:887]: IMHO it's fine to do here since it's related.

File: vhost-device-sound/src/audio_backends/alsa.rs (hunk 8 above)
Comment (stefano-garzarella) [reply vhost-device-sound/src/audio_backends/alsa.rs:199]: IMHO "request" is better.
Maybe you can do the rename in a next patch, but I'm not sure about it.

File: vhost-device-sound/src/device.rs [hunk 14]
```diff
@@ -443,77 +429,35 @@ impl VhostUserSoundThread {
...
+
+            let mem = atomic_mem.memory();
+
+            let mut reader = desc_chain
+                .clone()
+                .reader(&mem)
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            let in_header: VirtioSoundPcmXfer =
+                reader.read_obj().map_err(|_| Error::DescriptorNotFound)?;
```
Comment (stefano-garzarella) [inline vhost-device-sound/src/device.rs:441]: Why `Error::DescriptorNotFound` ?

File: vhost-device-sound/src/stream.rs [hunk 15]
```diff
@@ -239,85 +244,91 @@ impl Default for PcmParams {
...
-                addr.checked_add(offset)
-                    .expect("invalid guest memory address"),
-            )
-            .map_err(|_| Error::DescriptorReadFailed)?;
-        Ok(len as u32)
+        let mem = self.message.desc_chain.memory();
+
+        let Ok(mut reader) = self.message.desc_chain.clone().reader(mem) else {
+            return Err(Error::DescriptorReadFailed);
+        };
```
Comment (stefano-garzarella) [inline vhost-device-sound/src/stream.rs:280]: We use `map_err()` a lot in this crate, so I suggest continuing in that way:

```suggestion
        let mut reader = self
            .message
            .desc_chain
            .clone()
            .reader(mem)
            .map_err(|_| Error::DescriptorReadFailed)?;
```

And in other places.

File: vhost-device-sound/src/stream.rs (hunk 6 above)
Comment (stefano-garzarella) [reply vhost-device-sound/src/stream.rs:319]: I think `used_len` must be u32 for the spec, but a request can be bigger than `used_len`.
`used_len` tracks only the bytes written by the device, so IMHO is better to use `usize` and check for overflow when we set `used_len`.

File: vhost-device-sound/src/stream.rs (hunk 7 above)
Comment (stefano-garzarella) [reply vhost-device-sound/src/stream.rs:378]: yeah, I agree that we should check overflows

**Review by stefano-garzarella — COMMENTED** [review]
Great cleanup! I think we should also release a new version of virtio-queue ASAP since this PR looks in a good shape (I can do next week).

I left some minor comments.

File: vhost-device-sound/src/device.rs (hunk 12 above)
Comment (MatiasVara) [reply vhost-device-sound/src/device.rs:162]: I agree. However, the `UnexpectedWriteOnlyDescriptor` error requires as parameter the index of the descriptor that should not be `WriteOnly` but there is not such information. `reader` returns always `InvalidChain` in case or error. Should I keep `DescriptorReadFailed`?

File: vhost-device-sound/src/device.rs (hunk 14 above)
Comment (MatiasVara) [reply vhost-device-sound/src/device.rs:441]: I do not know. I think it should be `Error::DescriptorReadFailed`.

File: vhost-device-sound/src/device.rs (hunk 12 above)
Comment (stefano-garzarella) [reply vhost-device-sound/src/device.rs:162]: After this PR `UnexpectedWriteOnlyDescriptor` is not used in any place, so you can adapt it or remove it if you prefer to use `DescriptorReadFailed`. I don't have a strong opinion, so it's up to you ;-)

Comment (MatiasVara) [reply vhost-device-sound/src/device.rs:162]: is it OK if I add a new commit to remove the unused errors?

Comment (stefano-garzarella) [reply vhost-device-sound/src/device.rs:162]: Yep, sure.

File: vhost-device-sound/src/stream.rs (hunk 7 above)
Comment (MatiasVara) [reply vhost-device-sound/src/stream.rs:378]: This is part of the tests in which overflow is not possible for the moment. Do you think I should check for overflow anyway?

File: vhost-device-sound/src/stream.rs (hunk 6 above)
Comment (MatiasVara) [reply vhost-device-sound/src/stream.rs:319]: I'll use `usize` for `self.len()` and then check for overflow when getting `used_len`.

File: vhost-device-sound/src/stream.rs (hunk 7 above)
Comment (MatiasVara) [reply vhost-device-sound/src/stream.rs:378]: What would you think about this?
```suggestion
            (std::mem::size_of::<R>() as u32).saturating_add(payload_len),
```

Comment (stefano-garzarella) [reply vhost-device-sound/src/stream.rs:378]: Nope, maybe something more like `(std::mem::size_of::<R>() as u32).checked_add(payload_len).unwrpa()`, so it will panic if the overflow will happen.

Because a `Descriptor` can contain only a buffer where the length is on 32 bit.

File: vhost-device-sound/src/device.rs (hunk 13 above)
Comment (MatiasVara) [reply vhost-device-sound/src/device.rs:370]: Do youo mean something like this?:
```rust
            if u32::try_from(used_len).is_err() {
                log::error!("used_len has overflowed");
            }
```

Comment (stefano-garzarella) [reply vhost-device-sound/src/device.rs:370]: Yep, but at that point should we at least pass `u32::MAX` to `add_used()`?

I'm thinking about something like this (maybe better to use `warn` if we don't propagate the error to the guest):

```
let used_len = match u32::try_from(used_len) {
    Ok(len) => len,
    Err(len) => {
        log::warn!("used_len {} overflows u32", len);
        u32::MAX
    }
};
```

Comment (MatiasVara) [reply vhost-device-sound/src/device.rs:370]: Yes, we should pass `u32:MAX` in that case. I will add that, thanks.

Comment (MatiasVara) [reply vhost-device-sound/src/device.rs:370]: I wonder if this should not go in another new commit just before the adding of the use of Reader/Writer.

Comment (stefano-garzarella) [reply vhost-device-sound/src/device.rs:370]: Yeah, it's pre-existing, so good point.

Comment (MatiasVara) [reply vhost-device-sound/src/device.rs:370]: I just realized that this is partially pre-existing. In the previous code, `used_len` is u32 and no casting is required (Although the previous code does not check for overflow during additions). When we start to use Reader/Writer, `used_len` becomes `usize` and casting is required. Overflow can happen during this conversation. I think it would be better to add this check to the commit that targets the control queue instead of using a new commit. If I create a new commit for this, I would have to change everything again when adding Reader/Writer to the control queue. What do you think?

Comment (MatiasVara) [reply vhost-device-sound/src/device.rs:370]: Thanks, sorry for the confusion, I will update the PR and we can discuss then.

File: vhost-device-sound/src/stream.rs [hunk 16]
```diff
@@ -239,85 +245,108 @@ impl Default for PcmParams {
...
+            .map_err(|_| Error::DescriptorReadFailed)?;
+
+        let mut write_content = writer
+            .split_at(self.pos)
+            .map_err(|_| Error::DescriptorReadFailed)?;
+
+        let bytes_written = write_content
+            .write(buf)
             .map_err(|_| Error::DescriptorWriteFailed)?;
-        self.pos += len;
```
Comment (MatiasVara) [inline vhost-device-sound/src/stream.rs:305]: This change may be wrong.

File: vhost-device-sound/src/audio_backends/pipewire.rs [hunk 17]
```diff
@@ -375,18 +375,16 @@ impl AudioBackend for PwBackend {
 
                                 let mut start = 0;
                                 while n_samples > 0 {
```
Comment (MatiasVara) [inline vhost-device-sound/src/audio_backends/pipewire.rs:377]: I think we can remove this. We shall only write a request, i.e., period_bytes but not more than that.

File: vhost-device-sound/src/stream.rs (hunk 16 above)
Comment (MatiasVara) [reply vhost-device-sound/src/stream.rs:305]: I reverted this change since it breaks alsa audio backend. This change did not break pipewire because we are incrementing in the audio backend too. I think the correct thing is to fix pipewire so we do not increment two times and we only increment during `write_input()`.

**Comment by vireshk** [comment]
@Ablu since you have reviewed this earlier, can you do it again please ?

**Comment by MatiasVara** [comment]
I still do not like that in https://github.com/MatiasVara/vhost-device/blob/b1fcfcacb0ad0c1594951b0bba022958c90b25ae/vhost-device-sound/src/lib.rs#L312, this PR uses a descriptor. Shall I address that in this PR or in a new one? I shall not use `reponse_descriptor`.

**Comment by MatiasVara** [comment]
> > I still do not like that in https://github.com/MatiasVara/vhost-device/blob/b1fcfcacb0ad0c1594951b0bba022958c90b25ae/vhost-device-sound/src/lib.rs#L312, this PR uses a descriptor. Shall I address that in this PR or in a new one? I shall not use `response_descriptor`.
> 
> How difficult do you think it is? The best thing would be to do it now, but if too difficult, we can merge the code as it is now and then do it later.

I agree the best would be to address that in this PR. I will update the PR soon.

File: vhost-device-sound/src/device.rs [hunk 18]
```diff
@@ -155,86 +154,58 @@ impl VhostUserSoundThread {
...
-            }
+            let mut reader = desc_chain
+                .clone()
+                .reader(&mem)
+                .map_err(|_| Error::DescriptorReadFailed)?;
 
-            let request = desc_chain
-                .memory()
-                .read_obj::<VirtioSoundHeader>(desc_request.addr())
+            let mut writter_status = desc_chain
```
Comment (dorindabassey) [inline vhost-device-sound/src/device.rs:164]: just some small typo: 
```suggestion
            let mut writer_status = desc_chain
```

File: vhost-device-sound/src/device.rs [hunk 19]
```diff
@@ -495,77 +437,35 @@ impl VhostUserSoundThread {
...
-                        // to it. When we copy just after the request is enqueued, the guest's
-                        // userspace may or may not have updated the buffer contents. Guest driver
-                        // simply moves buffers from the used ring to the available ring without
-                        // knowing whether the content has been updated. The device only reads the
-                        // buffer from guest memory when the audio engine requires it, which is
-                        // about after a period thus ensuring that the buffer is up-to-date.
-                        buffers.push(Buffer::new(*descriptor, Arc::clone(&message), direction));
-                    }
+
+            let mem = atomic_mem.memory();
```
Comment (dorindabassey) [inline vhost-device-sound/src/device.rs:441]: maybe it's better to keep `atomic_mem.memory()` outside the loop to avoid repetitive calls.

File: vhost-device-sound/src/device.rs (hunk 18 above)
Comment (MatiasVara) [reply vhost-device-sound/src/device.rs:164]: Thanks! this typo is in another places too. I'll fix them.

**Comment by MatiasVara** [comment]
I just updated the PR based on last comments. Feel free to review @stefano-garzarella @Ablu

**Review by stefano-garzarella — APPROVED** [review]

**Comment by stefano-garzarella** [comment]
@stsquad @vireshk @epilys please, can you take a look?

**Review by dorindabassey — APPROVED** [review]
LGTM

**Review by stsquad — APPROVED** [review]
